packages/mcp/src/proxy/mcp-proxy.ts

```diff
@@ -305,243 +349,654 @@ export class MCPProxy {
+      console.error(
+        `[proxy] Auto-reconnection failed for ${serverName}:`,
+        error,
+      );
+      logError('server:auto_reconnect_failed', error, { name: serverName });
+
+      // Re-schedule reconnection - the ReconnectionManager will handle backoff
+      const reconnectionManager = this.reconnectionManagers.get(serverName);
+      if (reconnectionManager) {
+        reconnectionManager.scheduleReconnection();
+      }
     }
   }
 
-  private notifyToolListChanged(reason: string): void {
-    void this._server.sendToolListChanged().catch((error) => {
-      const message = error instanceof Error ? error.message : String(error);
-      console.error(
-        `[proxy] Failed to send tools/list_changed (${reason}): ${message}`,
-      );
-      logError('tools:list_changed_failed', error, { reason });
+  /**
+   * Connect to a single target server
+   * Extracted from connectToTargetServers for reuse in reconnection logic
+   */
+  private async connectToSingleServer(
+    targetServer: TargetServer | TargetServerZod,
+  ): Promise<void> {
+    logEvent('info', 'server:connect_start', {
+      name: targetServer.name,
+      command: targetServer.command,
+      args: targetServer.args,
+      hasAuth: !!(targetServer as TargetServerZod).auth,
+      hasTransport: !!(targetServer as TargetServerZod).transport,
     });
-  }
 
-  getTargetServers() {
-    return {
-      connected: Array.from(this.serverState.connected),
-      disconnected: Array.from(this.serverState.disconnected),
+    const client = new Client({
+      name: `proxy-client-${targetServer.name}`,
+      version: '1.0.0',
+    });
+
+    const extendedServer = targetServer as TargetServerZod;
+    const legacyServer = targetServer as TargetServer;
+
+    const baseCommand =
+      legacyServer.command ??
+      (extendedServer.transport?.type === 'stdio'
+        ? extendedServer.transport.command
+        : undefined) ??
+      '';
+    const baseArgs =
+      legacyServer.args ??
+      (extendedServer.transport?.type === 'stdio'
+        ? extendedServer.transport.args
+        : undefined) ??
+      [];
+
+    const targetForEnv: TargetServer = {
+      name: targetServer.name,
+      command: baseCommand,
+      args: Array.isArray(baseArgs) ? baseArgs : [],
+      env: legacyServer.env,
+      secretProviders: legacyServer.secretProviders,
     };
-  }
 
-  async populateToolCaches(): Promise<void> {
-    for (const [serverName, client] of this._clients) {
-      await this.refreshServerTools(serverName, client);
+    const shouldResolveSecrets =
+      (this._config.defaultSecretProviders?.length ?? 0) > 0 ||
+      (legacyServer.secretProviders?.length ?? 0) > 0;
+
+    const resolvedEnv = shouldResolveSecrets
+      ? await resolveServerEnvironment(
+          targetForEnv,
+          this._config,
+          this._configPath,
+        )
+      : ({
+          ...process.env,
+          ...legacyServer.env,
```

Comment:
When `shouldResolveSecrets` is false we skip `resolveServerEnvironment(...)` and instead set `resolvedEnv = { ...process.env, ...legacyServer.env }`. That bypasses the filtered passthrough list in `DefaultEnvironmentResolver` (`packages/mcp/src/proxy/env.ts:16-43`) and forwards **every** host environment variable to the target server. For most deployments (no secret providers configured) this leaks credentials, API keys, etc., to every connected server. We should still funnel through the resolver (or otherwise filter) even when no secret providers are defined.